src/python/pants/cache/s3_artifact_cache.py

```diff
@@ -0,0 +1,178 @@
+# coding=utf-8
+# Copyright 2014 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
+                        unicode_literals, with_statement)
+
+import logging
+import os
+
+import boto3
+from botocore import exceptions
+from botocore.config import Config
+from botocore.vendored.requests import ConnectionError, Timeout
+from botocore.vendored.requests.packages.urllib3.exceptions import ClosedPoolError
+from pyjavaproperties import Properties
+from six.moves.urllib.parse import urlparse
+
+from pants.cache.artifact_cache import ArtifactCache, NonfatalArtifactCacheError, UnreadableArtifact
+
+
+logger = logging.getLogger(__name__)
+
+_NETWORK_ERRORS = [
+  ConnectionError, Timeout, ClosedPoolError,
+  exceptions.EndpointConnectionError, exceptions.ChecksumError
+]
+
+def _connect_to_s3(config_file, profile_name):
+  # Yeah, I know it's gross but it spams the logs without it:
+  boto3.set_stream_logger(name='boto3.resources', level=logging.WARN)
+  boto3.set_stream_logger(name='botocore', level=logging.WARN)
+
+  boto_kwargs = {}
+  if profile_name:
+    boto_kwargs['profile_name'] = profile_name
+
+  try:
+    with open(config_file, 'r') as f:
+      p = Properties()
+      p.load(f)
+
+      access_key = p.get('accessKey')
+      if access_key:
+        logger.debug('Reading access key from {0}'.format(config_file))
+        boto_kwargs['aws_access_key_id'] = access_key
+
+      secret_key = p.get('secretKey')
+      if secret_key:
+        logger.debug('Reading access key from {0}'.format(config_file))
+        boto_kwargs['aws_secret_access_key'] = secret_key
+  except IOError:
+    logger.debug('Could not load {0}, using ENV vars'.format(config_file))
+
+  session = boto3.Session(**boto_kwargs)
+  config = Config(connect_timeout=4, read_timeout=4)
+  return session.resource('s3', config=config)
+
+
+_READ_SIZE_BYTES = 4 * 1024 * 1024
```

Comment:
Should this be an option? Or is it just super-standard to read 4MB chunks from S3?